src/Fcf/Data/Symbol.hs

```diff
@@ -0,0 +1,244 @@
+{-# LANGUAGE DataKinds              #-}
+{-# LANGUAGE PolyKinds              #-}
+{-# LANGUAGE TypeFamilies           #-}
+{-# LANGUAGE TypeInType             #-}
+{-# LANGUAGE TypeOperators          #-}
+{-# LANGUAGE UndecidableInstances   #-}
+{-# OPTIONS_GHC -Wall                       #-}
+{-# OPTIONS_GHC -Werror=incomplete-patterns #-}
+
+{-|
+Module      : Fcf.Data.Symbol
+Description : Type level symbols
+
+= Fcf.Data.Symbol
+
+Type-level symbols and functions for them.
+
+Note that the operators from this module conflict with "GHC.TypeLits".
+
+-}
+
+--------------------------------------------------------------------------------
+
+module Fcf.Data.Symbol
+    ( -- * Reexported type
+      -- | From "GHC.TypeList".
+
+      Symbol
+
+      -- * Functions
+
+    , Append
+    , Intercalate
+    , IsSpace
+    , IsNewLine
+    , IsTab
+    , IsSpaceDelim
+    , IsDigit
+
+     -- * Comparison functions
+
+    , SymbolOrd
+    , type (<=)
+    , type (>=)
+    , type (<)
+    , type (>)
+    , type (==)
+    )
+  where
+
+--------------------------------------------------------------------------------
+        
+import           GHC.TypeLits (Symbol)
+import qualified GHC.TypeLits as TL
+
+import           Fcf.Core (Eval, Exp)
+import           Fcf.Data.List (Foldr, Elem)
+import           Fcf.Data.Bool (type (||))
+import           Fcf.Utils (TyEq)
+
+import           Fcf.Combinators (type (=<<))
+
+--------------------------------------------------------------------------------
+
+-- | Append two type-level symbols.
+--
+-- === __Example__
+-- 
+-- >>> :kind! Eval (Append "hmm" " ok")
+-- Eval (Append "hmm" " ok") :: Symbol
+-- = "hmm ok"
+data Append :: Symbol -> Symbol -> Exp Symbol
+type instance Eval (Append s1 s2) = TL.AppendSymbol s1 s2
+
+
+-- | Intercalate type-level symbols.
```

Comment:
These functions seem a bit much. Have you actually used such functions at the type-level?

Maybe we should limit ourselves to what's already in base, and anything beyond should be a separate package. But I'm not 100% decided about this.

Reply:
I have only used it in one example: https://github.com/gspia/fcf-containers/blob/master/examples/Haiku.hs

(The language in the Haiku there is Finnish, sorry for that, the code and comments are English.)

Anyhow, my personal thought & motivation here goes roughly along the lines that once there is enough general purpose libs, it is much easier to actually build something. 

And I'm thinking like actually wanting to calculate things on compile-time, if the inputs and algorithm are already known and given. And to be able to compute, I need "String/Text" equivalent. My "Text"-module would be much more simpler if GHC had 

`toSymbols :: Symbol -> [Symbol]` function that would give ["a","b","c"] on "abc" input. I don't know if there is technical or theoretical difficulties with that kind of function but somehow I feel that if "abc" is known at compile-time, then also that list is.

So, these functions are just helpers to help implementing the Text-module. And in a way I feel that they are primitive (something comparable to Char-type), which is the reason I thought that they might could be in this lib. 

Anyhow, I'll remove them from this PR now.